### ISSUE TITLE
List.map calling mapper with modified list instead of original list

### ISSUE THREAD

**Issue by veerabio** [immutable-js/immutable-js#1649]
### What happened

map<M>(mapper: (value: T, key: number, iter: this) => M, context?: any): List<M>

**_iter_** should be the original list not the modified list.

### How to reproduce

`
var x = Immutable.fromJS([1,2,3,4]);

x.map((value, key, iter) => { console.log(iter.toJS()); return 'N';});
`

![image](https://user-images.githubusercontent.com/29557253/48641878-09932200-e990-11e8-8960-a598153ff405.png)

Seems to be broken after this
https://github.com/facebook/immutable-js/pull/1455

**Comment by mishawakerman** [comment]
We just spent a lot of time debugging an issue that appears to be the same as this:

```
const { List } = require('immutable');

const things = List([1,2,3,4]);

const things2 = things.map((thing, idx, others) => {
	let prev = others.get(idx - 1);
	return prev;
});

console.log(JSON.stringify(things2))
// things2 = [4, 4, 4, 4]
// should be [4,1,2,3] (as with immutable@3.8.2)
```

Notably, it appears to be caused by accessing (not modifying) the `others` argument and if you replace the line with `let prev = things.get(idx - 1);` then it works as expected.

I get that `v4` is still a release candidate but it's also the default version that's installed when you just run `yarn add immutable`.

**Comment by jdeniau** [comment]
Thank you for your bug report. The immutable-js oss fork will soon be merged and a fix for this issue has been included in #1833. Once this PR is merged, this issue will be resolved in the main branch. We will then do our best to to release the `4.0.0` version.

**Comment by jdeniau** [comment]
Hi!
The 4.0.0-rc.14 version has been released and should fix this issue.
Feel free to test it and re-open this issue if the issue is still present.
Thank you!
